### PR TITLE
Use alignment to store `Gc`'s rooted flag

### PR DESCRIPTION
**Pull request by apasel422**
Supplants #17

### CODE REVIEW

File: gc/src/lib.rs [hunk 1]
```diff
@@ -34,9 +36,7 @@ pub use gc::force_collect;
 ///
 /// See the [module level documentation](./) for more details.
 pub struct Gc<T: Trace + ?Sized + 'static> {
-    // XXX We can probably take advantage of alignment to store this
-    root: Cell<bool>,
-    ptr: Shared<GcBox<T>>,
+    ptr_root: Cell<NonZero<*const GcBox<T>>>,
```
Comment (mystor) [inline gc/src/lib.rs]: We should make sure to also have the PhantomData here that used to be in Shared<T>

File: gc/src/lib.rs [hunk 2]
```diff
@@ -60,19 +60,43 @@ impl<T: Trace> Gc<T> {
     /// let five = Gc::new(5);
     /// ```
     pub fn new(value: T) -> Self {
+        debug_assert!(mem::align_of::<GcBox<T>>() > 1);
```
Comment (mystor) [inline gc/src/lib.rs]: Let's make this an assert!(). It will be optimized out of release builds anyway.

File: gc/src/lib.rs [hunk 3]
```diff
@@ -60,19 +60,43 @@ impl<T: Trace> Gc<T> {
...
             (**ptr).value().unroot();
-            Gc { ptr: ptr, root: Cell::new(true) }
+            let gc = Gc { ptr_root: Cell::new(NonZero::new(*ptr)) };
+            gc.set_root();
+            gc
         }
     }
 }
 
+unsafe fn ptr<T: ?Sized + Trace>(ptr: NonZero<*const GcBox<T>>) -> NonZero<*const GcBox<T>> {
```
Comment (mystor) [inline gc/src/lib.rs]: Can we have a doc comment on this? It's use is not obvious from its name. Perhaps `remove_root_bit` would be a better name?

File: gc/src/lib.rs [hunk 4]
```diff
@@ -4,16 +4,18 @@
 //! It is marked as non-sendable because the garbage collection only occurs
 //! thread-locally.
 
-#![feature(borrow_state, coerce_unsized, core_intrinsics, optin_builtin_traits, shared, unsize)]
+#![feature(borrow_state, coerce_unsized, core_intrinsics, nonzero, optin_builtin_traits, shared, unsize)]
```
Comment (mystor) [inline gc/src/lib.rs]: I don't think we need the shared feature anymore, do we?

**Comment by mystor** [comment]
I left a few in line comments. I'm very happy that we can have the efficient flag storage and coerce unsized now at the same time! 

Thanks for the patch :)

File: gc/src/lib.rs (hunk 4 above)
Comment (apasel422) [inline gc/src/lib.rs]: `Shared` is still used by `GcState`.

**Comment by apasel422** [comment]
Updated.

**Comment by mystor** [comment]
LGTM, Thanks for the patch

**Pull request merged**